File: edukate-backend/src/main/java/io/github/sanyavertolet/edukate/backend/entities/files/SubmissionFileKey.java
package io.github.sanyavertolet.edukate.backend.entities.files;

import com.fasterxml.jackson.annotation.JsonTypeName;
import lombok.Getter;
import org.springframework.data.annotation.TypeAlias;

@TypeAlias("submission")
@JsonTypeName("submission")
public class SubmissionFileKey extends FileKey {
    @Getter
    private final String userId;
    @Getter
    private final String problemId;
    @Getter
    private final String submissionId;

    private SubmissionFileKey(String userId, String problemId, String submissionId, String fileName) {
        super(fileName);
        this.userId = userId;
        this.problemId = problemId;
        this.submissionId = submissionId;
    }

    public static SubmissionFileKey of(String userId, String problemId, String submissionId, String fileName) {
        return new SubmissionFileKey(userId, problemId, submissionId, fileName);
    }

    private static String constructKey(String userId, String problemId, String submissionId, String fileName) {
        return String.format("users/%s/submissions/%s/%s/%s", userId, problemId, submissionId, fileName);
    }

    public static String prefix(String userId, String problemId, String submissionId) {
        return constructKey(userId, problemId, submissionId, "");
    }

    @Override
    public String toString() {
        return constructKey(userId, problemId, submissionId, fileName);
    }
}

File: edukate-storage/src/main/kotlin/io/github/sanyavertolet/edukate/storage/keys/SubmissionFileKey.kt
package io.github.sanyavertolet.edukate.storage.keys

import com.fasterxml.jackson.annotation.JsonTypeName

@JsonTypeName("submission")
data class SubmissionFileKey(val userId: Long, val problemId: Long, val submissionId: Long, override val fileName: String) :
    FileKey {
    override fun toString() = prefix(userId, problemId, submissionId) + fileName

    override fun type() = "submission"

    override fun owner() = userId

    companion object {
        fun prefix(userId: Long, problemId: Long, submissionId: Long) =
            "users/${userId.toString()}/submissions/${problemId.toString()}/${submissionId.toString()}/"
    }
}
